File: app/src/main/java/com/pitchai/flickrrecentphotobrowser/presentation/presenter/GridPhotoListPresenterImp.java
package com.pitchai.flickrrecentphotobrowser.presentation.presenter;

import com.pitchai.flickrrecentphotobrowser.data.RestApiService;
import com.pitchai.flickrrecentphotobrowser.data.photoinfo.Photo;
import com.pitchai.flickrrecentphotobrowser.data.photoinfo.PhotoInfo;
import com.pitchai.flickrrecentphotobrowser.data.photoinfo.Photos;
import com.pitchai.flickrrecentphotobrowser.domain.Interactor.DefaultObserver;
import com.pitchai.flickrrecentphotobrowser.domain.Interactor.GetPhotoInfoList;
import com.pitchai.flickrrecentphotobrowser.domain.executor.PostExecutionThread;
import com.pitchai.flickrrecentphotobrowser.domain.executor.ThreadExecutor;
import com.pitchai.flickrrecentphotobrowser.presentation.view.GridPhotoListView;

import java.util.ArrayList;

/**
 * Created by pitchairajamani on 2/27/17.
 */

public class GridPhotoListPresenterImp implements Presenter {

    GridPhotoListView gridPhotoListView;
    private final GetPhotoInfoList getPhotoInfoList;

    public GridPhotoListPresenterImp(RestApiService apiService, ThreadExecutor executor, PostExecutionThread uiThread,
                                     GridPhotoListView gridPhotoListView) {
        this.gridPhotoListView = gridPhotoListView;
        getPhotoInfoList = new GetPhotoInfoList(apiService,executor,uiThread);
    }

    public void execute(int currentPageNumber) {
        getPhotoInfoList.execute(new PhotoInfoListObserver(), GetPhotoInfoList.Params.forPhotoList(currentPageNumber));
        gridPhotoListView.showLoading();
    }

    @Override
    public void resume() {

    }

    @Override
    public void pause() {

    }

    @Override
    public void destroy() {
        getPhotoInfoList.dispose();
        gridPhotoListView = null;
    }

    private  final class PhotoInfoListObserver extends DefaultObserver<PhotoInfo> {
        @Override public void onNext(PhotoInfo photoInfo) {
            // no-op by default.
            ArrayList<Photo> photoArrayList = null;
            int currentPageNumber = 0;
            if (photoInfo != null) {
                Photos photos = photoInfo.getPhotos();
                int mTotalPage = photos.getPages();
                if (photos != null && mTotalPage > 0) {
                    currentPageNumber = photos.getPage();
                    photoArrayList = new ArrayList<>();
                    Photo[] photoList = photos.getPhoto();
                    for (Photo photo : photoList) {
                        photoArrayList.add(photo);
                    }
                }
            }
            gridPhotoListView.hideLoading();
            gridPhotoListView.renderPhotoList(photoArrayList,currentPageNumber);
        }

        @Override public void onComplete() {
            // no-op by default.
        }

        @Override public void onError(Throwable exception) {
            // no-op by default.
            gridPhotoListView.hideLoading();
            gridPhotoListView.showError(exception.getMessage());
        }
    }
}

File: app/src/main/java/com/pitchai/flickrrecentphotobrowser/presentation/presenter/GridPhotoListPresenterImp.kt
package com.pitchai.flickrrecentphotobrowser.presentation.presenter

import com.pitchai.flickrrecentphotobrowser.data.FlickrRestApiService
import com.pitchai.flickrrecentphotobrowser.data.photoinfo.Photo
import com.pitchai.flickrrecentphotobrowser.data.photoinfo.PhotoInfo
import com.pitchai.flickrrecentphotobrowser.domain.Interactor.DefaultObserver
import com.pitchai.flickrrecentphotobrowser.domain.Interactor.GetPhotoInfoList
import com.pitchai.flickrrecentphotobrowser.domain.executor.PostExecutionThread
import com.pitchai.flickrrecentphotobrowser.domain.executor.ThreadExecutor
import com.pitchai.flickrrecentphotobrowser.presentation.view.GridPhotoListView
import java.util.*

/**
 * Created by pitchairajamani on 2/27/17.
 */
class GridPhotoListPresenterImp(apiService: FlickrRestApiService, executor: ThreadExecutor, uiThread: PostExecutionThread) : Presenter {

    var gridPhotoListView1 : GridPhotoListView ? = null

    private var getPhotoInfoList: GetPhotoInfoList ? = null

    fun setGridPhotoListView(view: GridPhotoListView?) {
        gridPhotoListView1 = view!!
    }

    fun execute(currentPageNumber: Int) {
        getPhotoInfoList?.execute(PhotoInfoListObserver(), GetPhotoInfoList.Params.Companion.forPhotoList(currentPageNumber))
        gridPhotoListView1?.showLoading()
    }

    override fun resume() {}
    override fun pause() {}
    override fun destroy() {
        getPhotoInfoList?.dispose()
        gridPhotoListView1 = null
    }

    private inner class PhotoInfoListObserver : DefaultObserver<PhotoInfo?>() {
        override fun onNext(photoInfo: PhotoInfo?) {
            // no-op by default.
            var photoArrayList: ArrayList<Photo?>? = null
            var currentPageNumber = 0
            if (photoInfo != null) {
                val photos = photoInfo.photos
                val mTotalPage = photos?.pages ?: 0
                if (photos != null && mTotalPage > 0) {
                    currentPageNumber = photos.page
                    photoArrayList = ArrayList()
                    val photoList = photos.photo
                    for (photo in photoList) {
                        photoArrayList.add(photo)
                    }
                }
            }
            gridPhotoListView1?.hideLoading()
            gridPhotoListView1?.renderPhotoList(photoArrayList, currentPageNumber)
        }

        override fun onComplete() {
            // no-op by default.
        }

        override fun onError(exception: Throwable) {
            // no-op by default.
            gridPhotoListView1?.hideLoading()
            gridPhotoListView1?.showError(exception.message)
        }
    }

    init {
        getPhotoInfoList = GetPhotoInfoList(apiService, executor, uiThread)
    }
}